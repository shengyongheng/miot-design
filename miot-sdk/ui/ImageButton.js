/**
 * @export
 * @module miot/ui/ImageButton
 * @description 按钮
 *
 */
import React from 'react';
import { Image, TouchableWithoutFeedback } from 'react-native';
import PropTypes from 'prop-types';
import { AccessibilityPropTypes, AccessibilityRoles, getAccessibilityConfig } from '../utils/accessibility-helper';
import { referenceReport } from '../decorator/ReportDecorator';
export default class ImageButton extends React.Component {
  constructor(props) {
    super(props);
    referenceReport('ImageButton');
    this.state = {
      buttonPressed: false
    };
  }
  static initialState = {
    buttonPressed: false
  };
  static propTypes = {
    source: PropTypes.any,
    highlightedSource: PropTypes.any,
    onPress: PropTypes.func,
    disabled: PropTypes.bool,
    style: PropTypes.any,
    accessible: AccessibilityPropTypes.accessible,
    accessibilityLabel: AccessibilityPropTypes.accessibilityLabel,
    accessibilityHint: AccessibilityPropTypes.accessibilityHint,
    accessibilityRole: AccessibilityPropTypes.accessibilityRole
  };
  static defaultProps = {
    source: null,
    highlightedSource: null,
    onPress: null
  };
  _buttonPressIn() {
    this.setState({ buttonPressed: true });
  }
  _buttonPressOut() {
    this.setState({ buttonPressed: false });
  }
  _isButtonPressed() {
    return this.state.buttonPressed;
  }
  render() {
    let { source, highlightedSource, disabled, onPress, accessible, accessibilityLabel, accessibilityHint, accessibilityRole, ...rest } = this.props;
    if (this._isButtonPressed() && highlightedSource) {
      source = highlightedSource;
    }
    return (
      <TouchableWithoutFeedback
        disabled={ disabled }
        onPress={ onPress }
        onPressIn={ this._buttonPressIn.bind(this) }
        onPressOut={ this._buttonPressOut.bind(this) }
        { ...getAccessibilityConfig({
          accessible: accessible,
          accessibilityRole: accessibilityRole || AccessibilityRoles.button,
          accessibilityLabel: accessibilityLabel,
          accessibilityHint: accessibilityHint,
          accessibilityState: {
            disabled: !!disabled
          }
        }) }
      >
        <Image
          source={ source }
          { ...rest }/>
      </TouchableWithoutFeedback>
    );
  }
}