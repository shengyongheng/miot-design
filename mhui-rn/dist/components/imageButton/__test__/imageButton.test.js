
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ImageButton from '../ImageButton';

configure({ adapter: new Adapter() });

describe('ImageButton', () => {
  describe('render', () => {
    it('render ImageButton', () => {
      const tree = renderer.create(
        <ImageButton
          style={{ width: 30, height: 30 }}
          source={{ uri: 'https://reactnative.dev/img/tiny_logo.png' }}
        />,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Interaction', () => {
    describe('onPress', () => {
      it('should call buttonPressIn', () => {
        const mockOnPress = jest.fn();
        const instance1 = shallow(<ImageButton
          style={{ width: 30, height: 30 }}
          source={{ uri: 'https://reactnative.dev/img/tiny_logo.png' }}
          onPress={mockOnPress}
        />).instance();
        instance1.buttonPressIn();
        instance1.handlePress();
        expect(mockOnPress).toHaveBeenCalled();
        expect(mockOnPress).toHaveBeenCalledTimes(1);
        expect(instance1.state.imageSource.uri).toBe('https://reactnative.dev/img/tiny_logo.png');
      });
      it('should call buttonPressIn and change image source', () => {
        const mockOnPress = jest.fn();
        const instance1 = shallow(<ImageButton
          style={{ width: 30, height: 30 }}
          source={{ uri: 'https://reactnative.dev/img/tiny_logo.png' }}
          highlightedSource={{ uri: 'https://reactnative.dev/img/header_logo.svg' }}
          onPress={mockOnPress}
        />).instance();
        instance1.buttonPressIn();
        instance1.handlePress();
        expect(mockOnPress).toHaveBeenCalled();
        expect(mockOnPress).toHaveBeenCalledTimes(1);
        expect(instance1.state.imageSource.uri).toBe('https://reactnative.dev/img/header_logo.svg');
        instance1.buttonPressOut();
        expect(instance1.state.imageSource.uri).toBe('https://reactnative.dev/img/tiny_logo.png');
      });
    });
  });
});
