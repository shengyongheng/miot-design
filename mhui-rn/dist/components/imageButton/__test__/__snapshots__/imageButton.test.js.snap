// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageButton render render ImageButton 1`] = `
<Image
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  source={
    Object {
      "uri": "https://reactnative.dev/img/tiny_logo.png",
    }
  }
  style={
    Object {
      "height": 30,
      "width": 30,
    }
  }
/>
`;
